File: plugins/sulis-execution/sdk/typescript/src/resources/findings.ts
import { kwargsToParams, resultPayload } from '../helpers.js';
import type {
  AsyncSubprocessTransport,
  SubprocessTransport,
  TransportConfig,
} from '../transport.js';
import type {
  FindingsDraftRemediationResult,
  FindingsRegisterResult,
} from '../types.js';

const BINARY = 'wpx-findings';

function common(config: TransportConfig): Record<string, unknown> {
  return { project: config.project, repo_root: config.repoRoot };
}

export interface FindingsRegisterParams {
  wp: string;
  severity: 'CRITICAL' | 'CONCERN' | 'ADVISORY';
  summary: string;
  file: string;
  evidence_json?: string;
  suggested_fix?: string;
  primitive?: string;
}

export interface FindingsDraftRemediationParams {
  source_finding: string;
  source_wp: string;
  auto_wp_id: string;
  severity: 'CONCERN' | 'ADVISORY';
  primitive?: 'Secure' | 'Harden' | 'Instrument' | 'Gate';
}

export class FindingsResource {
  constructor(
    private readonly transport: SubprocessTransport,
    private readonly config: TransportConfig,
  ) {}

  register(params: FindingsRegisterParams): FindingsRegisterResult {
    return resultPayload(
      this.transport.invoke(BINARY, 'register', {
        ...common(this.config),
        ...kwargsToParams(params),
      }),
    ) as unknown as FindingsRegisterResult;
  }

  draftRemediation(params: FindingsDraftRemediationParams): FindingsDraftRemediationResult {
    return resultPayload(
      this.transport.invoke(BINARY, 'auto-draft-wp', {
        ...common(this.config),
        primitive: 'Secure',
        ...params,
      }),
    ) as unknown as FindingsDraftRemediationResult;
  }
}

export class AsyncFindingsResource {
  constructor(
    private readonly transport: AsyncSubprocessTransport,
    private readonly config: TransportConfig,
  ) {}

  async register(params: FindingsRegisterParams): Promise<FindingsRegisterResult> {
    return resultPayload(
      await this.transport.invoke(BINARY, 'register', {
        ...common(this.config),
        ...kwargsToParams(params),
      }),
    ) as unknown as FindingsRegisterResult;
  }

  async draftRemediation(
    params: FindingsDraftRemediationParams,
  ): Promise<FindingsDraftRemediationResult> {
    return resultPayload(
      await this.transport.invoke(BINARY, 'auto-draft-wp', {
        ...common(this.config),
        primitive: 'Secure',
        ...params,
      }),
    ) as unknown as FindingsDraftRemediationResult;
  }
}

File: plugins/sulis-execution/sdk/python/sulis_execution/resources/findings.py
"""Findings resource — wraps `wpx-findings` subcommands."""
from __future__ import annotations

from typing import Optional

from sulis_execution._helpers import _kwargs_to_params, _result_payload
from sulis_execution.transport import (
    AsyncSubprocessTransport,
    SubprocessTransport,
    TransportConfig,
)
from sulis_execution.types import FindingsAutoDraftWpResult, FindingsRegisterResult

BINARY = "wpx-findings"


def _common(config: TransportConfig) -> dict:
    return {"project": config.project, "repo_root": str(config.repo_root)}


class FindingsResource:
    def __init__(
        self, transport: SubprocessTransport, config: TransportConfig
    ) -> None:
        self._transport = transport
        self._config = config

    def register(
        self,
        *,
        wp: str,
        severity: str,
        summary: str,
        file: str,
        evidence_json: Optional[str] = None,
        suggested_fix: Optional[str] = None,
        primitive: Optional[str] = None,
    ) -> FindingsRegisterResult:
        params = _kwargs_to_params({
            **_common(self._config),
            "wp": wp, "severity": severity, "summary": summary, "file": file,
            "evidence_json": evidence_json, "suggested_fix": suggested_fix,
            "primitive": primitive,
        })
        envelope = self._transport.invoke(BINARY, "register", params)
        return FindingsRegisterResult.model_validate(_result_payload(envelope))

    def auto_draft_wp(
        self,
        *,
        source_finding: str,
        source_wp: str,
        auto_wp_id: str,
        severity: str,
        primitive: str = "Secure",
    ) -> FindingsAutoDraftWpResult:
        params = {
            **_common(self._config),
            "source_finding": source_finding, "source_wp": source_wp,
            "auto_wp_id": auto_wp_id, "severity": severity, "primitive": primitive,
        }
        envelope = self._transport.invoke(BINARY, "auto-draft-wp", params)
        return FindingsAutoDraftWpResult.model_validate(_result_payload(envelope))


class AsyncFindingsResource:
    def __init__(
        self, transport: AsyncSubprocessTransport, config: TransportConfig
    ) -> None:
        self._transport = transport
        self._config = config

    async def register(
        self,
        *,
        wp: str,
        severity: str,
        summary: str,
        file: str,
        evidence_json: Optional[str] = None,
        suggested_fix: Optional[str] = None,
        primitive: Optional[str] = None,
    ) -> FindingsRegisterResult:
        params = _kwargs_to_params({
            **_common(self._config),
            "wp": wp, "severity": severity, "summary": summary, "file": file,
            "evidence_json": evidence_json, "suggested_fix": suggested_fix,
            "primitive": primitive,
        })
        envelope = await self._transport.invoke(BINARY, "register", params)
        return FindingsRegisterResult.model_validate(_result_payload(envelope))

    async def auto_draft_wp(
        self,
        *,
        source_finding: str,
        source_wp: str,
        auto_wp_id: str,
        severity: str,
        primitive: str = "Secure",
    ) -> FindingsAutoDraftWpResult:
        params = {
            **_common(self._config),
            "source_finding": source_finding, "source_wp": source_wp,
            "auto_wp_id": auto_wp_id, "severity": severity, "primitive": primitive,
        }
        envelope = await self._transport.invoke(BINARY, "auto-draft-wp", params)
        return FindingsAutoDraftWpResult.model_validate(_result_payload(envelope))
